addons/mail/models/mail_scheduled_message.py

```diff
@@ -54,6 +54,9 @@ class MailScheduledMessage(models.Model):
     # notify parameters (email_from, mail_server_id, force_email_lang,...)
     notification_parameters = fields.Text('Notification parameters')
 
+    # context used when posting the message to trigger some actions (eg. change so state when sending quotation)
+    send_context = fields.Text('Sending Context')
```

Comment:
I'm not exactly sure what keys are in the `send_context`, but won't this be error prone in a case like this:
schedule -> upgrade -> try to send with obsolete context ?
I guess the best way to handle that is to be sure that mails that are not sent because of context keys errors are reported as such to the author, in order for them to be able to reschedule ? Not sure of the feasibility though^^

Also, it seems there exists a `fields.Json`, isn't it more appropriate than `fields.Text` ?

Reply:
I was also a little bit sceptical about this approach (not only for upgrades) and thought of adding a list of (dis)allowed keys at some point.
The thing is that this change probably fixes some other issues we might be unaware of (for example, it's through this change that I noticed that the autofollow feature was missing when sending a scheduled message).

I don't think that "obsolete" context would be that much of an issue though as we usually check if a key is in the context to do something. So at worse, if for example the "mark_so_as_sent" key is renamed, the message will be sent but the SO state won't be changed. Which is annoying, but not catastrophic. But might be more annoying for some other cases, I don't really know.

However there might be some other flaws of storing the context. For example, the "mail_notify_force_send" key is set to False in the context when a message is sent using a cron. But with the proposed change, this key could be updated if it was stored in the send_context. (note that this one is easily avoidable by adding the force_send parameter when posing a scheduled message).

Maybe we should add the list of allowed keys (similarly to the one existing for the notification_parameters field).
But in this case we might not fix unknown issues. However we might suppose that if they are unknown, they are not that important, and the list can still be modified later if one is spotted.
I liked the approach because it was kind of "global", but It's probably better to have a "less risky" approach and fix other issues later (supposing that there will be other issues).

Otherwise we can completely change the approach and add boolean fields on the model such as "autofollow_on_post", "change_so_state_on_send" which will be set at create time based on the context.
This is probably the safest approach (will also remove the issue of renamed context keys).

What do you think?


Concerning the Json field, I didn't use it because it's not widely used and because the notification_parameters field on the same model uses a text field to store a json. But I can change it ofc, might be a better fit after all.

Reply:
Let's keep the context for now and see the opinion of the reviewer, since it's cleaned it does not contain field values, so like we discussed at worst we don't do something if there is an unknown key, not that big of a deal.

Thanks for the explanation!

For the Json field, I'd use it, it's there, it better matches what we do with it, and if we shouldn't do it for some reason I'd like for someone to tell us :3

Reply:
I would have used a text field and use ast.literal_eval in a try / except, but I suppose the Json will do the same. Using "clean_context" is a good step. Not sure what are the keys we are going to have in real life (lang probably, tz, stuff like that) ?

Adding options is something to do if we go for options on the composer itself. It is still not the case, as we don't have a lot of use cases (globally, autofollow, so / invoice / po auto confirm, maybe some stuff in pos I recently spotted). Could be a model-based set of "options", but I think the context is good enough for the time being.

Reply:
Concerning obsolete context: well that is not going to be a big issue, maybe in some rare cases some optional updates are going to be skipped. We should not worry on that I think, especially we don't often change context keys for the composer side actions.